File: codewars/6kyu/longest-palindrome/dduck-longest-palindrome.py
# longest-palindrome
# https://www.codewars.com/kata/54bb6f887e5a80180900046b

from unittest import TestCase


def get_longest_gap(s, leftIndex, rightIndex, max_gap):
    while leftIndex >= 0 and rightIndex < len(s) and s[leftIndex] == s[rightIndex]:
        leftIndex -= 1
        rightIndex += 1

    return max(max_gap, rightIndex - leftIndex - 1)


def longest_palindrome(s):
    index = 0
    max_gap = 0

    while index < len(s):
        max_gap = max(max_gap, get_longest_gap(s, index, index, max_gap), get_longest_gap(s, index, index + 1, max_gap))
        index += 1

    return max_gap


TestCase().assertEqual(longest_palindrome("a"), 1)
TestCase().assertEqual(longest_palindrome("aa"), 2)
TestCase().assertEqual(longest_palindrome("baa"), 2)
TestCase().assertEqual(longest_palindrome("aab"), 2)
TestCase().assertEqual(longest_palindrome("abcdefghba"), 1)
TestCase().assertEqual(longest_palindrome("baablkj12345432133d"), 9)

File: codewars/6kyu/longest-palindrome/dduck-longest-palindrome.js
// longest-palindrome
// https://www.codewars.com/kata/54bb6f887e5a80180900046b

const getLongestGap = (s, leftIndex, rightIndex, max) => {
  while (s[leftIndex] && s[leftIndex] === s[rightIndex]) {
    leftIndex -= 1;
    rightIndex += 1;
  }
  return Math.max(max, rightIndex - leftIndex - 1);
};

const longestPalindrome = (s, index = 0, max = 0) => {
  if (index >= s.length) {
    return max;
  }
  
  max = Math.max(max, getLongestGap(s, index, index, max), getLongestGap(s, index, index + 1, max));
  
  return longestPalindrome(s, index + 1, max);
};

console.log(longestPalindrome("a"), 1);
console.log(longestPalindrome("aa"), 2);
console.log(longestPalindrome("baa"), 2);
console.log(longestPalindrome("aab"), 2);
console.log(longestPalindrome("zyabyz"), 6, "Are you sure that is a palindrome?");
console.log(longestPalindrome("baabcd"), 4);
console.log(longestPalindrome("baablkj12345432133d"), 9);
